File: design-patterns/factory_design_pattern/vehicle_factory.cpp
#include "vehicle_factory.hpp"

Vehicle* VehicleFactory::getVehicle(string vehicleType){
    Vehicle* vehicle = nullptr;

    if(vehicleType == "Car"){
        vehicle = new Car();
    }
    else if(vehicleType == "Bike"){
        vehicle = new Bike();
    }
    else{
        cout<<"No matching vehicle."<<endl;
    }

    return vehicle;
}

File: design-patterns/CreationalDesignPatterns/FactoryDesignPattern/VehicleFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FactoryDesignPattern
{
    internal class VehicleFactory
    {
        public static IVehicle? GetVehicle(string? vehicleType)
        {
            IVehicle? vehicle = null;

            if(vehicleType == "Truck")
            {
                vehicle = new Truck();
            }
            else if(vehicleType == "Ship")
            {
                vehicle = new Ship();
            }

            return vehicle;
        }
    }
}

/*
 * If the if-else ladder becomes to big in GetVehicle method then we can create different concrete factory classes like TruckFactory, CarFactory
 * ShipFactory, etc.
 * And IFactory will be the interface for all the concrete factory classes with methods of CreateVehicle which will be implemented by them.
 * If the client wants Truck, IFactory vehicle = TruckFactory.CreateVehicle();
 * This may start converting to Abstract Factory Design Pattern.
 * 
 */
